web/web.go

```diff
@@ -0,0 +1,694 @@
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/gorilla/handlers"
+	_ "github.com/lib/pq"
+)
+
+// TODO(searce):
+// 1) Test cases for APIs
+// 2) API for saving/updating table-level changes.
+// 3) API for showing logs
+// 4) Split all routing to an route.go file
+// 5) API for downloading the schema file, ddl file and summary report file.
+// 6) Update schema conv after setting global datatypes and return conv. (setTypeMap)
+// 7) Add rateConversion() in schema conversion, ddl and report APIs.
+// 8) Add an overview in summary report API
+var mysqlTypeMap = make(map[string][]typeIssue)
+var postgresTypeMap = make(map[string][]typeIssue)
+
+func homeLink(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Welcome to Harbourbridge!")
+	w.WriteHeader(http.StatusOK)
+}
+
+// Driver config is used for direct database connection.
+type DriverConfig struct {
+	Driver   string `json:"Driver"`
+	Host     string `json:"Host"`
+	Port     string `json:"Port"`
+	Database string `json:"Database"`
+	User     string `json:"User"`
+	Password string `json:"Password"`
+}
+
+func databaseConnection(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var config DriverConfig
+	err = json.Unmarshal(reqBody, &config)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	var dataSourceName string
+	switch config.Driver {
+	case "postgres":
+		dataSourceName = fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable", config.Host, config.Port, config.User, config.Password, config.Database)
+	case "mysql":
+		dataSourceName = fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", config.User, config.Password, config.Host, config.Port, config.Database)
+	default:
+		http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", config.Driver), http.StatusBadRequest)
+		return
+	}
+	sourceDB, err := sql.Open(config.Driver, dataSourceName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("SQL connection error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	// Open doesn't open a connection. Validate database connection.
+	err = sourceDB.Ping()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Connection Error: %v. Check Configuration again.", err), http.StatusInternalServerError)
+		return
+	}
+	app.sourceDB = sourceDB
+	app.dbName = config.Database
+	app.driver = config.Driver
+	w.WriteHeader(http.StatusOK)
+}
+
+func convertSchemaSQL(w http.ResponseWriter, r *http.Request) {
+	if app.sourceDB == nil || app.dbName == "" || app.driver == "" {
+		http.Error(w, fmt.Sprintf("Database is not configured or Database connection is lost. Please set configuration and connect to database."), http.StatusNotFound)
+		return
+	}
+	conv := internal.MakeConv()
+	var err error
```

Comment:
Shouldn't we just call into conversion.ProcessInfoSchema here since it's the identical logic/code?

Or is it important that we generate different http errors for the "driver not supported" vs other errors?
Wouldn't http.StatusNotFound be an ok http error for all of the error cases generated by ProcessInfoSchema?

Reply:
The main reason why this code section is rewritten is because we have to get db name from app struct and ProcessInfoSchema is getting db name from environment variable.

Reply:
Got it -- thanks!
It's just a few lines of duplicate code in any case.